examples/MicroserviceExample/WorkerService/RabbitMqConsumer.cs

```diff
@@ -0,0 +1,132 @@
+
+namespace WorkerService
+{
+    public class RabbitMqConsumer : BackgroundService
+    {
+        private const string QueueName = "TestQueue";
+
+        private readonly ILogger<RabbitMqConsumer> logger;
+
+        private IConnection connection;
+        private IModel channel;
+        private ActivitySource activitySource;
+        private ITextFormat textFormat;
+        private TracerProvider tracerProvider;
+
+        public RabbitMqConsumer(ILogger<RabbitMqConsumer> logger)
+        {
+            this.logger = logger;
+        }
+
+        public override Task StartAsync(CancellationToken cancellationToken)
+        {
+            this.tracerProvider = Sdk.CreateTracerProvider((builder) =>
+            {
+                builder
+                    .AddActivitySource(nameof(RabbitMqConsumer))
+                    .UseZipkinExporter(b =>
+                    {
+                        var zipkinHostName = Environment.GetEnvironmentVariable("ZIPKIN_HOSTNAME") ?? "localhost";
+                        b.ServiceName = nameof(RabbitMqConsumer);
+                        b.Endpoint = new Uri($"http://{zipkinHostName}:9411/api/v2/spans");
+                    });
+            });
+
+            var connectionFactory = new ConnectionFactory()
+            {
+                HostName = Environment.GetEnvironmentVariable("RABBIT_HOSTNAME") ?? "localhost",
+                UserName = "guest",
+                Password = "guest",
+                Port = 5672,
+                DispatchConsumersAsync = true,
+                RequestedConnectionTimeout = TimeSpan.FromMilliseconds(3000),
+            };
+
+            this.connection = connectionFactory.CreateConnection();
+            this.channel = this.connection.CreateModel();
+            this.channel.QueueDeclarePassive(QueueName);
+            this.channel.BasicQos(0, 1, false);
+
+            this.activitySource = new ActivitySource(nameof(RabbitMqConsumer));
+
+            this.textFormat = new TraceContextFormat();
+
+            return base.StartAsync(cancellationToken);
+        }
+
+        public override async Task StopAsync(CancellationToken cancellationToken)
+        {
+            await base.StopAsync(cancellationToken);
+            this.tracerProvider.Dispose();
+            this.connection.Dispose();
+            this.channel.Dispose();
+        }
+
+        protected override async Task ExecuteAsync(CancellationToken stoppingToken)
+        {
+            stoppingToken.ThrowIfCancellationRequested();
+
+            var consumer = new AsyncEventingBasicConsumer(channel);
+            consumer.Received += async (bc, ea) =>
+            {
+                var parentContext = this.textFormat.Extract(ea.BasicProperties, ExtractTraceContextFromBasicProperties);
+
+                string activityName = $"{nameof(RabbitMqConsumer)}.{nameof(ExecuteAsync)}";
+                using (var activity = activitySource.StartActivity(activityName, ActivityKind.Server, parentContext))
```

Comment:
@reyang Being an expert on the spec. Should the span/activity created when consuming messages use the propagated context as parent or add as a link ("follows-from")? I would really like us to provide spec-correct guidance where possible.

Reply:
Short answer: I think the spec is not clear on this yet.

Long answer:

"follows-from" is an OpenTracing concept, which hasn't been formalized in the OpenTelemetry specification:
* https://github.com/open-telemetry/opentelemetry-specification/issues/562
* https://github.com/open-telemetry/opentelemetry-specification/issues/65

Most of the existing backends don't have support for links and I bet they already pass the same trace id over messaging system. And it looks like that the [W3C TraceContext for MQTT](https://github.com/w3c/trace-context-mqtt) is going to the same thing. I do think this could expose issue for batch processing and retry, using links instead of following the parent context seem to have more flexibility/power.

Reply:
Thanks, this is great context. For the immediate purposes I think it makes sense to settle on a parent/child relationship since that's the only thing that most backends support at the moment - if, in particular, Zipkin and/or Jaeger introduce support for visualizing links, then I think it could be cool to demonstrate that.

That aside, I am curious how the spec conversations will evolve. My gut says that there will be value in both use cases. That is, a parent/child relationship may make sense some times and links other times. Though that could make configuration and especially auto-instrumentation complex.